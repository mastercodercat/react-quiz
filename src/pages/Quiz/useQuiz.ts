import { useMemo, useEffect } from "react";
import { useHistory, useParams } from "react-router-dom";

import { RootState } from "store/store";
import { useAppSelector, useAppDispatch } from "store/hooks";
import { setAnswer } from "store/quiz";

import { RouteLink } from "Routes";

const useQuiz = () => {
  const dispatch = useAppDispatch();
  const history = useHistory();

  const { id } = useParams<{ id?: string }>();
  const { total, questions, answers } = useAppSelector(
    (state: RootState) => state.quiz
  );

  const currentId = useMemo(() => (id ? parseInt(id) : 0), [id]);

  useEffect(() => {
    if (total === 0) {
      history.push(RouteLink.Home);
      return;
    }
    if (currentId < 0 || currentId > total) {
      history.push(RouteLink.Score);
    }
  }, [currentId, total, questions, history]);

  const question = questions[currentId - 1];
  const answer = answers[currentId - 1];

  const handleAnswer = (answer: string) => {
    dispatch(setAnswer({ currentId: currentId - 1, answer }));
  };

  const handleNext = () => {
    history.push(`/quiz/${currentId + 1}`);
  };

  const handleExit = () => {
    history.push(RouteLink.Home);
  };

  return {
    currentId,
    total,
    question,
    answer,
    handleAnswer,
    handleNext,
    handleExit,
  };
};

export default useQuiz;
