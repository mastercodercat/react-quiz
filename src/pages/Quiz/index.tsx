import { Button } from "react-bootstrap";

import Layout from "components/Layout";
import Answer from "components/Answer";
import Question from "components/Question";

import useQuiz from "./useQuiz";

const Quiz = () => {
  const {
    currentId,
    question,
    answer,
    total,
    handleAnswer,
    handleNext,
    handleExit,
  } = useQuiz();

  if (!question) {
    return null;
  }

  return (
    <Layout title={question.category}>
      <Question question={question.question} />
      <Answer answer={answer} onAnswer={handleAnswer} />
      <h6 className="my-5">
        Question {currentId}/{total}
      </h6>
      <div className="d-flex justify-content-between">
        <Button variant="danger" className="px-4 py-2" onClick={handleExit}>
          Exit
        </Button>
        <Button variant="primary" className="px-4 py-2" onClick={handleNext}>
          Next
        </Button>
      </div>
    </Layout>
  );
};

export default Quiz;
