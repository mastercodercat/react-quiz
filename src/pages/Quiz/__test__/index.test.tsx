import React from "react";
import {
  screen,
  waitFor,
  act,
  RenderResult,
  fireEvent,
} from "@testing-library/react";

import renderTest, { store, mockDispatch } from "test/helpers/renderTest";
import Quiz from "..";
import { results as questions } from "test/mocks/questions.fixture.json";

let wrapper: RenderResult;

describe("Quiz", () => {
  beforeEach(() => {
    jest.useFakeTimers();
    wrapper = renderTest(<Quiz />, { routes: ["/quiz/1"] }, "/quiz/:id");
  });

  afterEach(() => {
    act(() => {
      jest.runOnlyPendingTimers();
    });
    jest.useRealTimers();
  });

  it("should test snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should display question", () => {
    expect(screen.getByText("Question 1/10")).toBeInTheDocument();

    expect(screen.queryByText(questions[0].category)).toBeInTheDocument();
    expect(screen.queryByText(questions[0].question)).toBeInTheDocument();
  });

  it("should display answer choices", () => {
    expect(screen.getByTestId("true")).toBeInTheDocument();
    expect(screen.getByTestId("false")).toBeInTheDocument();
  });

  it("should be able to select answer", async () => {
    await waitFor(() => {
      expect(screen.getByTestId("true")).toBeInTheDocument();
    });

    fireEvent.click(screen.getByTestId("true") as Element);

    await waitFor(() => {
      expect(mockDispatch).toHaveBeenCalled();
    });
  });

  it("should navigate to next question", async () => {
    fireEvent.click(screen.getByText("Next") as Element);

    expect(screen.queryByText(questions[1].category)).toBeInTheDocument();
    expect(screen.queryByText(questions[1].question)).toBeInTheDocument();
  });
});
