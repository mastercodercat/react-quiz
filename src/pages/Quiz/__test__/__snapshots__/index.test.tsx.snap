// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz should test snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row"
      >
        <div
          class="my-5 mx-auto p-5 bg-secondary bg-opacity-25 col-lg-6 col-md-8 col-sm-12"
        >
          <div
            class="mb-5 row"
          >
            <div
              class="text-center col-sm-12"
            >
              <h1>
                General Knowledge
              </h1>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-sm-12"
            >
              <div>
                This is the correct spelling of Supercalifragilisticexpialidocious.
              </div>
              <form
                class="mt-4"
              >
                <div
                  class="form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    data-testid="true"
                    name="answer"
                    type="radio"
                  />
                  <label
                    class="form-check-label"
                    title=""
                  >
                    True
                  </label>
                </div>
                <div
                  class="form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    data-testid="false"
                    name="answer"
                    type="radio"
                  />
                  <label
                    class="form-check-label"
                    title=""
                  >
                    False
                  </label>
                </div>
              </form>
              <h6
                class="my-5"
              >
                Question 
                1
                /
                10
              </h6>
              <div
                class="d-flex justify-content-between"
              >
                <button
                  class="px-4 py-2 btn btn-danger"
                  type="button"
                >
                  Exit
                </button>
                <button
                  class="px-4 py-2 btn btn-primary"
                  type="button"
                >
                  Next
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row"
    >
      <div
        class="my-5 mx-auto p-5 bg-secondary bg-opacity-25 col-lg-6 col-md-8 col-sm-12"
      >
        <div
          class="mb-5 row"
        >
          <div
            class="text-center col-sm-12"
          >
            <h1>
              General Knowledge
            </h1>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-sm-12"
          >
            <div>
              This is the correct spelling of Supercalifragilisticexpialidocious.
            </div>
            <form
              class="mt-4"
            >
              <div
                class="form-check form-check-inline"
              >
                <input
                  class="form-check-input"
                  data-testid="true"
                  name="answer"
                  type="radio"
                />
                <label
                  class="form-check-label"
                  title=""
                >
                  True
                </label>
              </div>
              <div
                class="form-check form-check-inline"
              >
                <input
                  class="form-check-input"
                  data-testid="false"
                  name="answer"
                  type="radio"
                />
                <label
                  class="form-check-label"
                  title=""
                >
                  False
                </label>
              </div>
            </form>
            <h6
              class="my-5"
            >
              Question 
              1
              /
              10
            </h6>
            <div
              class="d-flex justify-content-between"
            >
              <button
                class="px-4 py-2 btn btn-danger"
                type="button"
              >
                Exit
              </button>
              <button
                class="px-4 py-2 btn btn-primary"
                type="button"
              >
                Next
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
